man/Repository/cmdlog.man
.so ../macros
.TH cmdlog Repository %G% "BitMover, Inc." BitKeeper
.\"    =================== BitKeeper Command Log ===================
.SH NAME
bk cmd_log \- show the log of commands executed in this repository
.SH SYNOPSIS
.B bk cmdlog
.br
.B bk cmdlog \-a
.SH DESCRIPTION
.LP
There are two command logs maintained by BitKeeper, both in the
BitKeeper/log directory: repo_log and cmd_log.  The first logs only
commands which operate at the repository level, specifically:
.RS .5i
clone
.br
commit
.br
export
.br
pull
.br
push
.RE
The second logs all other commands.
To see the repository commands, run
.AX
bk cmdlog
.XA
To see all commands, run
.AX
bk cmdlog -a
.XA
Commands are listed in least recent to most recent order.


RELEASE-NOTES
============================================================================
Release notes for BitKeeper version 6.0.2 (released Oct 29, 2013)

    This is a bugfix release (one feature snuck in).

Bugfixes

  bk diffs
    - Fixed a bug where 'bk diffs -NL' did not list extra files.
    - bk diffs -s was not working because of the switch to -u as
      the default.
    - Function header matching in 'bk diffs -p' now works with C++
      names.

  bk difftool
    - When looking at diffs over different types of files, a binary
      file in the list would cause diffs for the next text file to not
      be displayed.

  bk gone
    - People would try and run 'bk gone foo.c' which is incorrect, the
      correct usage is:
	bk gone `bk log -r+ -nd:ROOTKEY:`
      bk gone now works harder to validate that the arguments passed
      are actually keys.

  Other bugfixes
    - Fix a bug in the BK6->BK5 compat code where some files were not
      converted correctly to the BK5 format.
    - Stale locks on windows were not always being cleaned immediately.
    - Ignore extra files in .bk which were causing problems in Mac OS.

Features

   bk annotate
    - MD5 keys are now supported in the annotation options by running
      either 'bk annotate -a5' or 'bk annotate -A5'.

============================================================================
Release notes for BitKeeper version 6.0.1 (released Aug 9, 2013)

    This is a bugfix release (a few features snuck in).

Bugfixes

  bk clone
    - 'bk clone -@' wasn't populating comps when the remote repository
      was a subset of the local baseline.
    - 'bk clone -r<rev> [--upgrade|--downgrade] repo localcopy' could
      fail in some cases, and has been fixed.
    - 'bk clone -r<rev> bk://host/repo localcopy' will now fail if the
      BK client version is bk-6.0 and the bkd is bk-6.0.1 or later.
      The solution is to upgrade the client.

  bk commit
    - 'bk commit -f' was still prompting for comments.
    - 'bk commit' had lost some status reporting; that has been
      restored.

  bk collapse
    - 'bk collapse' would leave components marked as pending; this caused
      pull to error with a "KEY MISMATCH" message.
    - 'bk fix/collapse' in a gate was only partially disabled, it would
      collapse in the product but would refuse to collapse in
      components.

  bk diffs
    - 'bk -Ucx diffs -N' would only print diffs for one new file;
      fixed to print diffs for all new files

  bk difftool
    - The file listing box on difftool resizes itself to be as large
      as the longest pathname shown (it used to truncate).
    - 'bk difftool -S -Lpath/to/component' is now supported.
      (-S == scan only the current component)
    - The 'next' button in bk difftool didn't come up highlighted when
      there was more than one file.
    - Better error reporting with -L when a single GCA doesn't exist.

  bk pull
    - A pull from a remote repository served by a bk version 5.x could fail
      if a component clone was part of the operation.  It works now.

  bk push
    - A push from a local repository served by a bk version 5.x to a remote
      repository that uses new storage format of bk-6 could fail if a component
      clone was part of the operation.  It works now.

  bk revtool
    - A particular sequence of clicks and launching difftool could
      make revtool unresponsive.

    - When launching revtool with a line number (bk revtoo +lno),
      the comments for the revision that added that line were not being shown.

  bk sfiles
    - In rare cases, a pending component would not be seen as pending
      ('bk -pA' did not list them but it does now).

  Misc
    - In rare cases, permissions in the .bk/SCCS directory could be set
      wrong and cause bk to fail.  The known case has been fixed and
      the problem is now repaired automatically.

    - In another permissions failure, when a user didn't have write
      permissions on the repository, commit would fail with an
      unhelpful error message.  The problem is now reported more
      clearly.

  Graphical tools
    - When resizing graphical tools quickly, the resize labels could
      get out of sync; they update more frequently now.

Features
    - bk difftool now supports -w/-b to ignore white space changes
      like 'bk diffs'.  There is toolbar control of this feature as
      well.

    - BitKeeper versions prior to 6.0 may now clone repositories in the new
      format; they will automatically be downgraded to a compatible format.

============================================================================
Release notes for BitKeeper version 6.0 (released Jul 05, 2013)

    The primary focus of the 6.0 effort is performance improvements.

New repository format for performance
    BK6 has a new file format designed for performance.

    The new repository format cannot be read by bk-5.x releases
    of BitKeeper, but clone can convert a new repository to a bk-5.x
    readable repository.

Compatibility
    New file formats are a cause for concern but we've done the extra
    work so you may try it without forcing everyone to upgrade. BK6 can
    pull/push the new format from/to bk5's format.

    To enable the the new, faster format:
  
	bk clone --upgrade bk5-repo bk6-repo
  
    To downgrade to be compatible with bk5 (or bk4 if it is a non-nested
    repo):

	bk clone --downgrade bk6-repo bk5-repo

    To allow clones/pulls/pushes from bk4 & bk5, serve the repositories
    with a bk6 bkd:

    	cd bk6-repo
	bk pull bk://bk6-server/bk5-repo
	bk push bk://bk6-server/bk5-repo
 
Performance improvements:
    In the examples below a non-nested test repository contained 55,000
    files with 282,000 changesets on an NFS server.  Both the server
    and client caches were dropped (made cold) before each operation.

    Note that none of the performance improvements depend on the use of
    a file system notifier.  We considered and rejected such an approach

src/cmdlog.c
/*
 * Copyright 2011-2013,2016 BitMover, Inc
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "system.h"
#include "sccs.h"
#include "range.h"
#include "nested.h"

typedef struct {
	u32	all:1;	      /* -a: show cmd_log instead of repo_log */
	u32	verbose:1;    /* -v: print raw logfile */
	u32	standalone:1; /* -S: standalone */
	u32	relative:1;   /* -r: relative */
	time_t	cutoff;       /* -c: cutoff date */
	char	*pattern;
	search	search;
	int	complen;      /* max component pathname (for -N) */
} Opts;

typedef struct {
	char	*prefix;	/* component owning log */
	FILE	*fp;		/* open to comp's cmd_log file */
	char	*buf;		/* next line from logfile */
	struct timeval tv;	/* timestamp on this line */
} logmux;

private void	parsePrint(Opts *opts, char *prefix, char *buf);
private void	parseTimestamp(logmux *lmp);
private int	logmux_oldest(char **lmlist);
private void	logmux_free(void *lmp);

int
cmdlog_main(int ac, char **av)
{
	Opts	*opts;
	FILE	*f;
	nested	*n;
	comp	*cp;
	int	c, i, nested;
	char	*t, *log;
	logmux	*lmp;
	char	**lmlist = 0;
	char	buf[MAXPATH];
	longopt	lopts[] = {
		{ "standalone", 'S' },
		{ 0, 0 }
	};

	opts = new(Opts);
	while ((c = getopt(ac, av, "Sac;rv", lopts)) != -1) {
		switch (c) {
		    case 'S': opts->standalone = 1; break;
		    case 'a': opts->all = 1; break;
		    case 'c': opts->cutoff = range_cutoff(optarg + 1); break;
		    case 'r':
			opts->relative = opts->verbose = opts->all = 1;
			break;
		    case 'v': opts->verbose = 1; break;
		    default: bk_badArg(c, av);
		}
	}
	nested = bk_nested2root(opts->standalone);
	if (av[optind]) {
		if (av[optind+1]) usage();
		if (av[optind][0] != '^' && !opts->all) {
			opts->pattern = aprintf("^%s/", av[optind]);
		} else {
			opts->pattern = aprintf("%s/", av[optind]);
		}
		opts->search = search_parse(opts->pattern);
	}

	log = opts->all ? "cmd_log" : "repo_log";

	/* standalone case */
	unless (nested) {
		concat_path(buf, proj_root(0), "/BitKeeper/log/");
		concat_path(buf, buf, log);
		unless (f = fopen(buf, "r")) return (1);
		while ((t = fgetline(f))) parsePrint(opts, 0, t);
		fclose(f);
		if (opts->pattern) free(opts->pattern);
		return (0);
	}

	n = nested_init(0, 0, 0, NESTED_PENDING);
	assert(n);
	EACH_STRUCT(n->comps, cp, i) {
		unless (C_PRESENT(cp)) continue;
		lmp = new(logmux);
		lmp->prefix = strdup(cp->path);
		c = strlen(cp->path);
		if (c > opts->complen) opts->complen = c;
		sprintf(buf, "%s/BitKeeper/log/%s", lmp->prefix, log);
		unless (exists(buf)) {
			lmp->fp = 0;
			logmux_free(lmp);
			continue;
		}
		unless (lmp->fp = fopen(buf, "r")) {
			perror("fopen");
			return (1);
		}
		/* grab the first line, if any */
		unless (lmp->buf = fgetline(lmp->fp)) {
			logmux_free(lmp);
			continue;
		}
		parseTimestamp(lmp);
		lmlist = addLine(lmlist, lmp);
	}
	nested_free(n);

	while ((i = logmux_oldest(lmlist))) {
		lmp = (logmux *)lmlist[i];
		parsePrint(opts, lmp->prefix, lmp->buf);
		unless (lmp->buf = fgetline(lmp->fp)) {
			removeLineN(lmlist, i, logmux_free);
			continue;
		}
		parseTimestamp(lmp);
	}
	free(opts);
	return (0);
}

private void
parsePrint(Opts *opts, char *prefix, char *buf)
{
	time_t	t;
	u32	usec = 0;
	char	*p, *p1, *equals;
	char	*version;
	char	*user;
	int	yelled = 0, ind, logver = 0;
	double	rmsec = 0.0;

	user = buf;
	for (p = log_versions; *p; ++p) {
		if (*p == buf[0]) {
			logver = p - log_versions;
			if (logver > LOGVER) {
				if (yelled) return;
				printf("cannot display this "
				    "log entry; please upgrade\n");
				yelled = 1;
				return;
			}
			user++;
			break;
		}
	}

	for (p = user; (*p != ' ') && (*p != '@'); p++);
	*p++ = 0;
	t = strtoul(p, &version, 10);
	while (isspace(*version)) ++version;
	if (logver > 0) {
		p = version + 1; // skip decimal point
		usec = strtoul(p, &version, 10);
		p = version;
		rmsec = strtod(p, &version);
		while (isspace(*version)) ++version;
	}
	if (*version == ':') {
		p = version;
		*p = 0;
		version = 0;
	} else {
		char *q;

		unless (p = strchr(p, ':')) return;
		*p = 0;

		unless (isalnum(*version)) {
			version = 0;
		} else  {
			for (q = 1+version; *q; ++q) {
				if ( (*q & 0x80) || (*q < ' ') ) {
					version = 0;
					break;
				}
			}
		}
	}
	unless (t >= opts->cutoff) return;
	p+=2; // skip colon and extra space inserted by log indenting
	chomp(p);
	/* find indent, if any */
	for (p1 = p; isspace(*p1); p1++);
	ind = p1 - p;

	/*
	 * if we got a pattern, do pattern matching
	 * skip this log entry on no match
	 */
	if (opts->pattern) {
		unless (search_either(p, opts->search)) return;
		/*
		 * if somebody is looking for foo, then
		 * they'll also match the "bk cmdlog foo" so
		 * don't print ourselves, unless we are
		 * actually looking for cmdlog entries
		 */
		unless (strneq(opts->pattern, "cmdlog", 6) ||
		    strneq(opts->pattern, "^cmdlog", 7)) {
			if (strneq(p, "cmdlog ", 7)) return;
		}
	}

	/*
	 * - the presence of a indent means this was not a user
	 *   run command.
	 * - if there is no equal sign it's not a command completion
	 * - if we have no verbose flag, we skip all but entries
	 *   for completed user commands
	 */
	equals = strstr(p, " = ");
	if (!opts->verbose && (!equals || ind)) return;

	/* output requested annotations */
	if (opts->verbose) {
		switch (logver) {
		    case 0:
			printf("%-8s %.19s %14s ",
			    user, ctime(&t), version ? version : "");
			break;
		    case 1:
			if (opts->relative) {
				printf("%.3fs ", rmsec);
			} else {
				printf("%-8s %.19s.%03u %14s ",
				    user, ctime(&t), usec / 1000,
				    version ? version : "");
			}
			break;
		    default:
			// Should never happen
			printf("Unknown log version ");
			break;
		}
		if (prefix) printf("[%-*s] ", opts->complen, prefix);
	}
	if (equals) *equals = 0;

	/*
	 * Output command.  If opts->verbose is set, do
	 * not prefix with "bk ".
	 * Note: the ind+1 is to balance the cmd_log being
	 * written with printf("%*s", x, " "); which
	 * prints out 1 space for x = 0, otherwise it prints x
	 * spaces.  Above (the +=2) we eat one space
	 * corresponding to x = 0, so x = 4 means 3 spaces,
	 * and x = 8 means 7 spaces.  So we're unmathing it and
	 * getting it to look regular.
	 */
	if (ind) printf("%*s", ind+1, " ");
	if (!opts->verbose && !strneq(p1, "bk ", 3)) {
		fputs("bk ", stdout);
	}
	fputs(p1, stdout);
	if (equals && opts->verbose) {
		*equals = ' ';
		printf("%s", equals);
	}
	printf("\n");
}

private void
parseTimestamp(logmux *lmp)
{
	char	*p, *p1;
	int	logver = 0;

	for (p = log_versions; *p; ++p) {
		if (*p == lmp->buf[0]) {
			logver = p - log_versions;
			break;
		}
	}
	for (p = lmp->buf; *p != ' '; p++);
	p++;
	lmp->tv.tv_sec = strtoul(p, &p1, 10);
	if (logver > 0) {
		lmp->tv.tv_usec = strtoul(++p1, 0, 10);
	} else {
		lmp->tv.tv_usec = 0;
	}
}

/*
 * Pick the nested cmd_log line to print from an array of the next line
 * from each component.
 * Lines are sorted by time.
 */
private int
logmux_oldest(char **lmlist)
{
	int	i, ret = 0;
	logmux	*lmp, *oldest = 0;

	EACH_STRUCT(lmlist, lmp, i) {
		if (!oldest ||
		    ((timercmp(&lmp->tv, &oldest->tv, ==)
			? (strcmp(lmp->prefix, oldest->prefix) < 0)
			: timercmp(&lmp->tv, &oldest->tv, <)))) {
			oldest = lmp;
			ret = i;
		}
	}

	return (ret);
}

private void
logmux_free(void *lmp)
{
	logmux	*p = lmp;

	free(p->prefix);
	if (p->fp) fclose(p->fp);
	free(p);
}
